docs/rfc/rfc-022-semi-permanent-testnet.md

```diff
@@ -0,0 +1,202 @@
+
+## Discussion
+
+The Discussion Section proposes a variety of aspects of maintaining a testnet for Tendermint.
+
+### Number of testnets
+
+There should probably be one testnet per maintained branch of Tendermint,
+i.e. one for the `main` branch
+and one per `v0.N.x` branch that the authors maintain.
+
+We may eventually discover that there is good reason to run more than one testnet for a branch,
+perhaps due to a significant configuration variation.
+
+### Testnet lifecycle
+
+The document has used the terms "long-lived" and "semi-permanent" somewhat interchangeably.
+The intent of the testnet being discussed in this RFC is to exist indefinitely;
+but there is a practical understanding that there will be testnet instances 
+which will be retired due to being associated with a no-longer-maintained release of Tendermint.
+Alternatively, we may discover a good reason to destroy and recreate the `main` testnet
+(such as if we build a new application for running the testnet).
+
+### Testnet MVP
+
+The minimum viable testnet meets this set of features:
+
+- The testnet self-updates following a new commit pushed to Tendermint's `main` branch on GitHub
+  (there are some omitted steps here, such as CI building appropriate binaries and
+  somehow notifying the testnet that a new build is available)
+- The testnet runs the Tendermint KV store for MVP
+- The testnet operators are notified if any node's process exits for any reason other than
+  a restart for a new binary
+- The testnet operators are notified if any node stops updating blocks,
+  and by extension if a chain halt occurs
+- The testnet has a minimum of 1 full node and 3 validators
+- The testnet has a reasonably low, constant throughput of transactions -- say 30 tx/min --
+  and the testnet operators are notified if that throughput drops below 75% of target
+  sustained over 5 minutes
+- The testnet only needs to run in a single datacenter/cloud-region for MVP
+- The testnet is running directly on VMs or compute instances;
+  while Kubernetes or other orchestration frameworks may offer many significant advantages,
+  the Tendermint engineers should not be required to learn those tools in order to
+  perform basic debugging
+
+### Testnet medium-term goals
+
+The medium-term goals are intended to be achievable within the 6-12 month time range
+following the launch of MVP.
+These goals could realistically be roadmapped following the launch of the MVP testnet.
+
+- The `main` testnet has more than 20 nodes
+- In addition to the `main` testnet,
+  there is at least one testnet associated with one release branch
+- The testnet no longer is simply running the Tendermint KV store;
+  now it is built on a more complex, custom application
+  that deliberately exercises a greater portion of the Tendermint stack
+- Each testnet is spread across at least two cloud providers,
+  in order to communicate over a network more closely resembling use of Tendermint in "real" chains
```

Comment:
Can it be two geographic zones of the same provider?

Reply:
Moreover, should we consider emulating WAN latencies between nodes as well, or this would be more suited for MVP?

Reply:
At the moment, I think it makes sense to plan on running in multiple providers for the medium-term plan -- partially because it will allow us to optimize for lower costs, but one nice side effect is that we likely will not run across any kind of "optimized routing" that the CSP may provide between its own datacenters.

IMO the WAN latency emulation partially falls into the chaos engineering concept a few lines down, but I think that is out of scope of a medium plan. It is probably representative of a real network to have some peers with very low network latency, but the majority of the peers having a relatively much higher latency.